File: vscode/src/services/ShareMessage.test.ts
import { describe, expect, it } from "vitest";
import { buildShareCopyMessage, buildShareEmail, buildSocialShareUrl } from "./ShareMessage.js";

const BASE = { branch: "feature/sidebar", url: "https://acme.jolli.ai/b/x", decisionCount: 4 };

describe("buildShareEmail", () => {
	it("teases the branch, decision count, and decision titles", () => {
		const { subject, body } = buildShareEmail({ ...BASE, titles: ["Add dark mode", "Cache results"] });
		expect(subject).toBe('How we built "feature/sidebar" — 4 decisions on Jolli Memory');
		expect(body).toContain("4 decisions");
		expect(body).toContain("• Add dark mode");
		expect(body).toContain("• Cache results");
		expect(body).toContain(BASE.url);
		expect(body).toContain("no login, no install");
	});

	it("singularizes one decision and omits the bullet list when there are no titles", () => {
		const { subject, body } = buildShareEmail({ ...BASE, decisionCount: 1, titles: [] });
		expect(subject).toContain("1 decision on Jolli Memory");
		expect(body).not.toContain("A few of the decisions inside:");
		expect(body).toContain(BASE.url);
	});
});

describe("buildShareCopyMessage", () => {
	it("includes up to two teaser titles and the link", () => {
		const msg = buildShareCopyMessage({ ...BASE, titles: ["Add dark mode", "Cache results", "Third"] });
		expect(msg).toContain("feature/sidebar");
		expect(msg).toContain("4 decisions");
		expect(msg).toContain("Add dark mode");
		expect(msg).toContain("Cache results");
		expect(msg).not.toContain("Third"); // capped at 2
		expect(msg).toContain(BASE.url);
	});

	it("drops the 'incl.' clause when there are no titles", () => {
		const msg = buildShareCopyMessage({ ...BASE, titles: [] });
		expect(msg).not.toContain("incl.");
		expect(msg).toContain("4 decisions");
		expect(msg).toContain(BASE.url);
	});
});

describe("buildSocialShareUrl", () => {
	const input = {
		branch: "feature/x",
		url: "https://acme.jolli.ai/b/abc",
		decisionCount: 4,
		titles: ["Drop Redux for signals", "Cache at the edge"],
	};
	const enc = encodeURIComponent(input.url);

	it("X copy is punchy and teases concrete decisions (no emoji crutch)", () => {
		const u = buildSocialShareUrl("x", input);
		expect(u).toContain("https://twitter.com/intent/tweet?text=");
		expect(u).toContain(`url=${enc}`);
		const text = decodeURIComponent(u);
		expect(text).toContain("Drop Redux for signals");
		expect(text).toContain("not just the diff");
		expect(text).not.toContain("👀");
	});

	it("LinkedIn carries only the URL (it renders from OG tags)", () => {
		expect(buildSocialShareUrl("linkedin", input)).toBe(
			`https://www.linkedin.com/sharing/share-offsite/?url=${enc}`,
		);
	});

	it("Reddit gets a title-style headline", () => {
		const u = buildSocialShareUrl("reddit", input);
		expect(u).toContain("https://www.reddit.com/submit?url=");
		expect(decodeURIComponent(u)).toContain("feature/x: 4 decisions");
	});

	it("WhatsApp copy is casual and embeds the URL", () => {
		const u = buildSocialShareUrl("whatsapp", input);
		expect(u).toContain("https://wa.me/?text=");
		expect(decodeURIComponent(u)).toContain(input.url);
		expect(decodeURIComponent(u)).toContain("How we built feature/x");
	});

	it("Telegram passes url + text separately", () => {
		const u = buildSocialShareUrl("telegram", input);
		expect(u).toContain("https://t.me/share/url?url=");
		expect(u).toContain(enc);
		expect(u).toContain("text=");
	});

	it("omits the 'incl.' teaser when there are no titles", () => {
		const u = buildSocialShareUrl("x", { ...input, titles: [] });
		expect(decodeURIComponent(u)).not.toContain("incl.");
	});

	it("commit-kind copy leads with 'A commit on' across platforms", () => {
		expect(decodeURIComponent(buildSocialShareUrl("x", { ...input, kind: "commit" }))).toContain(
			"A commit on feature/x",
		);
		expect(decodeURIComponent(buildSocialShareUrl("reddit", { ...input, kind: "commit" }))).toContain(
			"A commit on feature/x: 4 decisions",
		);
		expect(decodeURIComponent(buildSocialShareUrl("whatsapp", { ...input, kind: "commit" }))).toContain(
			"A commit on feature/x",
		);
		expect(decodeURIComponent(buildSocialShareUrl("telegram", { ...input, kind: "commit" }))).toContain(
			"A commit on feature/x",
		);
	});
});

describe("commit-kind email + copy message", () => {
	it("email leads with 'A commit on' and reasons about a commit", () => {
		const { subject, body } = buildShareEmail({ ...BASE, titles: ["Add dark mode"], kind: "commit" });
		expect(subject).toBe('A commit on "feature/sidebar" — 4 decisions on Jolli Memory');
		expect(body).toContain("reasoning behind a commit");
	});

	it("copy message leads with 'A commit on'", () => {
		const msg = buildShareCopyMessage({ ...BASE, titles: [], kind: "commit" });
		expect(msg).toContain("A commit on feature/sidebar");
	});
});

File: intellij/src/test/kotlin/ai/jolli/jollimemory/services/ShareMessageTest.kt
package ai.jolli.jollimemory.services

import ai.jolli.jollimemory.services.ShareMessage.ShareKind
import ai.jolli.jollimemory.services.ShareMessage.SocialPlatform
import io.kotest.matchers.shouldBe
import io.kotest.matchers.string.shouldContain
import io.kotest.matchers.string.shouldNotContain
import io.kotest.matchers.string.shouldStartWith
import org.junit.jupiter.api.Test

class ShareMessageTest {

    private val branch = "feature/login"
    private val url = "https://acme.jolli.ai/s/abc123"
    private val titles = listOf("Chose JWT over sessions", "Rate-limited the token endpoint")

    @Test
    fun `email subject leads with branch and decision count`() {
        val email = ShareMessage.buildShareEmail(ShareMessage.ShareMessageInput(branch, url, 3, titles))
        email.subject shouldBe "How we built \"$branch\" — 3 decisions on Jolli Memory"
        email.body shouldContain "A few of the decisions inside:"
        email.body shouldContain "  • Chose JWT over sessions"
        email.body shouldContain url
        email.body shouldContain "— Shared via Jolli Memory"
    }

    @Test
    fun `email singular decision has no plural s`() {
        val email = ShareMessage.buildShareEmail(ShareMessage.ShareMessageInput(branch, url, 1, emptyList()))
        email.subject shouldContain "1 decision on Jolli Memory"
        email.subject shouldNotContain "1 decisions"
        // No teaser block when there are no titles.
        email.body shouldNotContain "A few of the decisions inside:"
    }

    @Test
    fun `commit-kind email uses commit phrasing`() {
        val email = ShareMessage.buildShareEmail(
            ShareMessage.ShareMessageInput(branch, url, 2, titles, ShareKind.COMMIT),
        )
        email.subject shouldStartWith "A commit on \"$branch\""
        email.body shouldContain "the reasoning behind a commit on the \"$branch\""
    }

    @Test
    fun `copy message teases the first two decisions`() {
        val msg = ShareMessage.buildShareCopyMessage(ShareMessage.ShareMessageInput(branch, url, 5, titles))
        msg shouldContain "How we built $branch — 5 decisions incl. “Chose JWT over sessions” & “Rate-limited the token endpoint”"
        msg shouldContain url
    }

    @Test
    fun `X share url encodes text and carries the url`() {
        val out = ShareMessage.buildSocialShareUrl(
            SocialPlatform.X,
            ShareMessage.SocialShareInput(branch, url, 3, titles),
        )
        out shouldStartWith "https://twitter.com/intent/tweet?text="
        out shouldContain "url=https%3A%2F%2Facme.jolli.ai%2Fs%2Fabc123"
        // encodeURIComponent-style spaces (%20), never "+".
        out shouldNotContain "+"
    }

    @Test
    fun `linkedin share url is url-only`() {
        val out = ShareMessage.buildSocialShareUrl(
            SocialPlatform.LINKEDIN,
            ShareMessage.SocialShareInput(branch, url, 3, titles),
        )
        out shouldBe "https://www.linkedin.com/sharing/share-offsite/?url=https%3A%2F%2Facme.jolli.ai%2Fs%2Fabc123"
    }

    @Test
    fun `reddit share url carries a title param`() {
        val out = ShareMessage.buildSocialShareUrl(
            SocialPlatform.REDDIT,
            ShareMessage.SocialShareInput(branch, url, 3, titles),
        )
        out shouldStartWith "https://www.reddit.com/submit?url="
        out shouldContain "title="
        out shouldContain "read-only%20dev%20memory"
    }

    @Test
    fun `whatsapp and telegram share urls exist and differ`() {
        val wa = ShareMessage.buildSocialShareUrl(SocialPlatform.WHATSAPP, ShareMessage.SocialShareInput(branch, url, 1, emptyList()))
        val tg = ShareMessage.buildSocialShareUrl(SocialPlatform.TELEGRAM, ShareMessage.SocialShareInput(branch, url, 1, emptyList()))
        wa shouldStartWith "https://wa.me/?text="
        tg shouldStartWith "https://t.me/share/url?url="
    }
}
